File: backend/app/services/intelligence_pipeline.py
"""Analyse (OpenAI) then score (deterministic) — port of intelligencePipeline.js.

Runs the two SEPARATE modules per researched company. OpenAI output is
whitelisted (analyst-only); the score is owned by the ScoringEngine.
"""
import asyncio
from typing import Any, Callable, Dict, List, Optional

from ..logging_config import get_logger
from . import openai_service, scoring_engine

log = get_logger("pipeline")


def _researched(c: Dict[str, Any]) -> bool:
    return (c.get("ai") or {}).get("researchStatus") in ("done", "completed", "cached")


def _scored(c: Dict[str, Any]) -> bool:
    return bool((c.get("ai") or {}).get("score"))


async def run(
    companies: List[Dict[str, Any]],
    rescore: bool = True,
    limit: Optional[int] = None,
    pace: float = 0.0,
    save_cb: Optional[Callable[[], None]] = None,
    save_every: int = 5,
) -> Dict[str, Any]:
    """Analyse + score researched companies. Resumable + batch-friendly:
      • rescore=False → only companies not yet scored (so repeated batches advance),
      • ``limit`` → process at most N this call (one batch),
      • ``pace`` → seconds between LLM calls (smooths Groq free-tier rate limits),
      • ``save_cb`` → persist progress every ``save_every`` companies (no work lost).
    Returns scored/failed for this batch plus ``remaining`` researched-but-unscored."""
    targets = [c for c in companies if _researched(c) and (rescore or not _scored(c))]
    if limit is not None:
        targets = targets[:limit]
    scored = failed = 0
    for i, c in enumerate(targets):
        result = await openai_service.analyze(c["crm"]["brandName"], c["ai"].get("research") or {})
        if not result.get("ok"):
            c["ai"]["analysisError"] = result.get("error", "analysis failed")
            failed += 1
        else:
            intel = openai_service.whitelist(result["data"])
            c["ai"]["intelligence"] = intel
            c["ai"]["analysisError"] = None
            c["ai"]["industry"] = intel.get("industry")
            c["ai"]["financialHealth"] = intel.get("financialHealth")
            c["ai"]["advertisingActivity"] = intel.get("advertisingActivity")
            c["ai"]["growthSignals"] = intel.get("growthSignals")
            c["ai"]["expansionSignals"] = intel.get("expansionSignals")
            c["ai"]["aiSummary"] = intel.get("businessSummary")
            # SEPARATE deterministic scoring (app-owned)
            s = scoring_engine.score(c)
            c["ai"]["score"] = s
            c["ai"]["opportunityScore"] = s["total"]
            c["ai"]["recommendation"] = s["recommendation"]
            c["ai"]["reason"] = s["businessReason"]
            scored += 1
        if save_cb and (i + 1) % save_every == 0:
            save_cb()
        if pace and i < len(targets) - 1:
            await asyncio.sleep(pace)
    if save_cb:
        save_cb()
    remaining = sum(1 for c in companies if _researched(c) and not _scored(c))
    log.info("Analysis batch: %s scored, %s failed, %s remaining", scored, failed, remaining)
    return {"scored": scored, "failed": failed, "remaining": remaining, "batch": len(targets)}

File: src/services/intelligencePipeline.js
/**
 * ALIP · IntelligencePipeline (service — Phase 4 orchestration)
 * ---------------------------------------------------------------------------
 * Runs the two SEPARATE modules for every researched company, in order:
 *   1. AIIntelligenceService.analyze()  → structured intelligence (no scores)
 *   2. ScoringEngine.score()            → deterministic Opportunity Score
 *
 * The AI output is stored at company.ai.intelligence and mirrored into the flat
 * placeholder fields; the score (owned by the app) is stored at company.ai.score
 * plus opportunityScore / recommendation / reason. Only companies whose research
 * is done/cached are processed; skipped duplicates are ignored.
 *
 * `run` is provider-injectable and store-agnostic for testability.
 * Mirrors future backend: backend/services/intelligence_pipeline.py
 */
(function (root) {
  'use strict';

  var isNode = (typeof module !== 'undefined' && module.exports);
  function dep(name, path) { return isNode ? require(path) : (root.ALIP && root.ALIP[name]); }

  function isResearched(c) {
    return c.ai && (c.ai.researchStatus === 'done' || c.ai.researchStatus === 'cached');
  }

  /**
   * @param {Object[]} companies canonical companies (mutated: ai.intelligence, ai.score …)
   * @param {Object} opts { ai, scoring, onUpdate(companies), reScore }
   */
  function run(companies, opts) {
    opts = opts || {};
    var ai = opts.ai || dep('AIIntelligenceService', '../services/aiIntelligenceService.js');
    var scoring = opts.scoring || dep('ScoringEngine', '../domain/scoringEngine.js');
    var onUpdate = opts.onUpdate || function () {};

    var targets = companies.filter(function (c) {
      if (!isResearched(c)) { return false; }
      return opts.reScore ? true : !(c.ai.intelligence);   // skip already-analysed unless re-scoring
    });

    var chain = Promise.resolve();
    targets.forEach(function (company) {
      chain = chain
        .then(function () { return ai.analyze(company); })
        .then(function (intelligence) {
          // 1) store analyst intelligence (whitelisted — no decisions inside)
          company.ai.intelligence = intelligence;
          company.ai.industry = intelligence.industry;
          company.ai.financialHealth = intelligence.financialHealth;
          company.ai.advertisingActivity = intelligence.advertisingActivity;
          company.ai.growthSignals = intelligence.growthSignals;
          company.ai.expansionSignals = intelligence.expansionSignals;
          company.ai.aiSummary = intelligence.businessSummary;
          company.ai.analysisError = null;

          // 2) SEPARATE deterministic scoring (the app owns the number)
          var result = scoring.score(company);
          company.ai.score = result;
          company.ai.opportunityScore = result.total;
          company.ai.recommendation = result.recommendation;
          company.ai.reason = result.businessReason;

          onUpdate(companies);
        })
        .catch(function (err) {
          company.ai.analysisError = String(err && err.message ? err.message : err);
          onUpdate(companies);
        });
    });

    return chain.then(function () { return companies; });
  }

  /** Progress stats for the UI. */
  function stats(companies) {
    var researched = 0, scored = 0, failed = 0;
    (companies || []).forEach(function (c) {
      if (isResearched(c)) { researched++; }
      if (c.ai && c.ai.score) { scored++; }
      if (c.ai && c.ai.analysisError) { failed++; }
    });
    return { researched: researched, scored: scored, failed: failed, remaining: researched - scored - failed };
  }

  var IntelligencePipeline = { run: run, stats: stats, isResearched: isResearched };

  if (isNode) { module.exports = IntelligencePipeline; }
  else { root.ALIP = root.ALIP || {}; root.ALIP.IntelligencePipeline = IntelligencePipeline; }
})(typeof self !== 'undefined' ? self : this);
